lab_4/main.py

```diff
@@ -3,31 +3,91 @@
 """
 
 from ngrams.ngram_trie import NGramTrie
-
+import re
 
 def tokenize_by_sentence(text: str) -> tuple:
-    pass
+    if not isinstance(text, str):
+        raise ValueError
+    elif not text:
+        return ()
+
+    sentences_in_text = re.split('[\n.?!]', text)
+    new_text = []
+
+    for sentence in sentences_in_text:
+        if len(sentence.strip()):
+            new_sentences = re.sub('[^a-z \n]', '', sentence.lower()).split()
+            if len(new_sentences):
+                new_text.append(new_sentences + ['<END>'])
+
+    full_text = []
+    for new_sentence in new_text:
+        for word in new_sentence:
+            full_text.append(word)
+
+    return tuple(full_text)
 
 
 class WordStorage:
     def __init__(self):
-        pass
+        self.storage = {}
+        self.id = 1
 
     def _put_word(self, word: str):
-        pass
+
+        if not isinstance(word, str) \
+                or not len(word):
+            raise ValueError
+
+        if word not in self.storage:
+            self.storage[word] = self.id
+            self.id += 1
+
+        return self.storage[word]
 
     def get_id(self, word: str) -> int:
-        pass
+
+        if not isinstance(word, str):
+            raise ValueError
+
+        for word1 in self.storage:
```

Comment:
There is a much faster way to do so. Read about accessing the keys/values of a dictionary.